server/functions/framework/compiled_function.go

```diff
@@ -193,232 +201,334 @@ func (c *CompiledFunction) Eval(ctx *sql.Context, row sql.Row) (interface{}, err
 	// much more work, but there's a performance penalty for runtime overload resolution in Postgres as well.
 	if c.IsOperator {
-		return c.resolveOperator(parameters, sources)
+		return c.resolveOperator(argTypes, sources)
 	} else {
-		return c.resolveFunction(parameters, sources)
+		return c.resolveFunction(argTypes, sources)
 	}
 }
 
+type overloadMatch struct {
+	params overloadParamPermutation
+	casts  []TypeCastFunction
+}
+
 // resolveFunction resolves a function according to the rules defined by Postgres.
 // https://www.postgresql.org/docs/15/typeconv-func.html
-func (c *CompiledFunction) resolveFunction(parameters []pgtypes.DoltgresType, sources []Source) (*OverloadDeduction, []TypeCastFunction, error) {
-	// First we'll discard all overloads that have a different length, or that do not have implicitly-convertible params
-	var convertibles [][]pgtypes.DoltgresTypeBaseID
-	var casts [][]TypeCastFunction
-	for _, overload := range c.AllOverloads {
-		if len(overload) == len(parameters) {
-			isConvertible := true
-			overloadCasts := make([]TypeCastFunction, len(overload))
-			// Polymorphic parameters must be gathered so that we can later verify that they all have matching base types
-			var polymorphicParameters []pgtypes.DoltgresType
-			var polymorphicTargets []pgtypes.DoltgresType
-			for i, overloadParam := range overload {
-				if parameters[i].BaseID() == pgtypes.DoltgresTypeBaseID_Null {
-					overloadCasts[i] = identityCast
-				} else if polymorphicType, ok := overloadParam.GetRepresentativeType().(pgtypes.DoltgresPolymorphicType); ok && polymorphicType.IsValid(parameters[i]) {
-					overloadCasts[i] = identityCast
-					polymorphicParameters = append(polymorphicParameters, polymorphicType)
-					polymorphicTargets = append(polymorphicTargets, parameters[i])
-				} else if overloadCasts[i] = GetImplicitCast(parameters[i].BaseID(), overloadParam); overloadCasts[i] == nil {
-					if sources[i] == Source_Constant && parameters[i].BaseID().GetTypeCategory() == pgtypes.TypeCategory_StringTypes {
-						overloadCasts[i] = stringLiteralCast
-					} else {
-						isConvertible = false
-						break
-					}
-				}
-			}
-			if isConvertible && c.checkPolymorphicTypes(polymorphicParameters, polymorphicTargets) {
-				convertibles = append(convertibles, overload)
-				casts = append(casts, overloadCasts)
-			}
-		}
+func (c *CompiledFunction) resolveFunction(argTypes []pgtypes.DoltgresType, sources []Source) (FunctionInterface, overloadMatch, error) {
+
+	compatibleOverloads := c.typeCompatibleOverloads(argTypes, sources)
+	if len(compatibleOverloads) == 0 {
+		return nil, overloadMatch{}, nil
 	}
+
 	// If we've found exactly one match then we'll return that one
-	if len(convertibles) == 1 {
-		matchedOverload := c.Functions
-		for _, parameter := range convertibles[0] {
-			matchedOverload = matchedOverload.Parameter[parameter]
+	if len(compatibleOverloads) == 1 {
```

Comment:
Need to handle this case from the documentation:
> After such expansion the function might have effective argument types identical to some non-variadic function. In that case the function appearing earlier in the search path is used, or if the two functions are in the same schema, the non-variadic one is preferred.

* https://www.postgresql.org/docs/15/typeconv-func.html
Since everything is defined in the same schema for now, we just need to prefer non-variadic over variadic for multiple matches.

Reply:
Added a TODO for now since it can't happen yet, will return to this later.